src/service/AssignmentService.ts

```diff
@@ -108,11 +135,78 @@ export async function checkAnswer(db: DatabaseConnection, questionId: number, us
     // Cleanup the temporary database
     tempDb.destroy();
 
+
+
+    const answerSelectResult = await db.exec(`SELECT id, query_history FROM answers WHERE ua_id = @ua AND question_id = @question;`,
+        { ua: uaId, question: questionId });
+
+    checkQueryResult(answerSelectResult, false);
+    
+    if (answerSelectResult.rows.length === 0) {
+        // INSERT new answer
+        const answerInsertResult = await db.exec(`INSERT INTO answers (ua_id, question_id, answer, query_history, score)
+            VALUES (@ua, @question, @answer, @history, @score)`,
+        {
+            ua: uaId,
+            question: questionId,
+            answer: userQuery,
+            history: userQuery,
+            score: isCorrect ? question.points : 0
+        });
+
+        checkQueryResult(answerInsertResult, false);
+    } else {
+        // UPDATE existing answer
+        
+        const answerId = answerSelectResult.rows[0].id;
+        const history = answerSelectResult.rows[0].query_history.split(HISTORY_DELIMITOR);
+        history.push(userQuery);
+
+        const answerInsertResult = await db.exec(`UPDATE answers SET ua_id = @ua, question_id = @question, answer = @answer, query_history = @history, score = @score
+            WHERE id = @answerId`,
+        {
+            ua: uaId,
+            question: questionId,
+            answer: userQuery,
+            history: history.join(HISTORY_DELIMITOR),
+            score: isCorrect ? question.points : 0,
+            answerId: answerId
+        });
+
+        checkQueryResult(answerInsertResult, false);
+    }
+
     return {
         status: 200,
         responseJson: {
             success: isCorrect,
             result: actual.success ? actual.rows : actual.error
         }
     };
+}
+
+export async function clearHistory(db: DatabaseConnection, questionId: number): Promise<RouteResult> {
+    const uaResult = await db.exec(`
+        SELECT ua.id
+            FROM questions q
+            JOIN assignments a ON a.id = q.assignment_id
+            JOIN users_assignments ua ON ua.assignment_id = a.id AND ua.user_id = @user
+            WHERE q.id = @id`,
+        {id: questionId, user: DEFAULT_USER_ID});
+    
+    checkQueryResult(uaResult, true);
+
+    const uaId = uaResult.rows[0].id;
+
+    const result = await(db.exec(`
+        UPDATE answers SET query_history = ''
+        WHERE ua_id = @ua AND question_id = @question`,
```

Comment:
I'm not a fan of writing out SQL in these functions. I think the better apporach would be to use the repository pattern and handle the queries in a separate file that's sole job is to interact with the DB. SQL code should probably not appear in our AssignmentService, it should just be our business logic. Including data access details means we're violating the separation of concerns principle.

Ideally this clearHistory function for example would just look more like this:
clearHistory(...) { 
  const userAssignmentResult = await assignmentRepo.getAssignmentByQuestionId(questionId);
  checkQueryResults(userAssignmenetResult, true);
  const result = await assignmentRepo.setQueryHistory(questionId, uaId)
}

This is much better for other developers working on our code as we're working only at the abstraction level of the business rules. It's nice to only show what they need to know at this level, not to show them all the details of data access they probably don't care about. If they want those details, they can always click into the repo function.